cozy/codegen/cxx.py

```diff
@@ -278,7 +281,7 @@ def construct_concrete(self, t : Type, e : Exp, out : Exp):
             x = self.fv(t, "x")
             indices = range(len(t.ts))
             return SSeq(
-                SEscape("{indent}{rv} = {rhs};\n", ["rv", "rhs"], [x, e]),
+                SEscape("{indent}{t} {rv} = {rhs};\n", ["t", "rv", "rhs"], [t, x, e]),
```

Comment:
In languages like C and C++, you cannot always declare a variable by writing `Type varName`. For instance, an array of ints might be declared `int a[]`.

Codegen is set up to support this: the visitor for type also accepts a name parameter, and inserts the name where it belongs. You will need to write this line differently to use that feature:

```suggestion
                SEscape("{indent}" + self.visit(t, x) + " = {rhs};\n", ("rhs",), (e,)),
```

Doing it this way also means you do not need to alter `visit_TTuple` above.

Also, I removed the use of lists in favor of tuples. Lists are not hashable, which makes them dangerous. There are places in Cozy where we use syntax trees in sets and maps. It might not affect this line, but as a general rule I try to use only tuples in syntax tree objects.

Reply:
That's cool! Thanks a lot